### PR TITLE
Utilise expression context cache to store some more expensive expression calculation results

### PR DESCRIPTION
**Pull request by nyalldawson**
Speeds up some complex lengthy calculations over large layers

### CODE REVIEW

File: src/core/expression/qgsexpressionfunction.cpp [hunk 1]
```diff
@@ -3481,6 +3481,13 @@ static QVariant fcnGetFeature( const QVariantList &values, const QgsExpressionCo
   }
 
   const QVariant &attVal = values.at( 2 );
+
+  const QString cacheValueKey = QStringLiteral( "getfeature:%1:%2:%3" ).arg( vl->id() ).arg( attributeId ).arg( attVal.toString() );
```
Comment (m-kuhn) [inline src/core/expression/qgsexpressionfunction.cpp]: Wouldn't a single multiattribute `.arg` call be faster?

**Review by m-kuhn — APPROVED** [review]
Nice, that will hopefully help with some of the slower candidates.

File: src/core/expression/qgsexpressionfunction.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/expression/qgsexpressionfunction.cpp]: You can't when you have any non-string arguments

Comment (m-kuhn) [reply src/core/expression/qgsexpressionfunction.cpp]: `QString::number( fid )`?

**Pull request merged**